File: steps/mock-anthropic.mjs
// A local server that speaks the real Anthropic Messages API (POST /v1/messages,
// both blocking and SSE streaming) and replays a *scripted* scenario. The agent
// code is unmodified — only ANTHROPIC_BASE_URL points here — so there is no test
// branch in the teaching code.
//
// Which scripted turn to return is derived from the request itself: the number
// of assistant messages already in `messages` is the index of the next turn.
// That is stateless and matches however the client replays history.
//
// A scenario is { id, turns: [ turn, ... ] } where a turn is:
//   { "tools": [ { "name": "...", "input": {...} } ], "text": "optional preamble" }  -> stop_reason tool_use
//   { "text": "..." }                                                                 -> stop_reason end_turn
//   optional per-turn { "usage": { "input_tokens": N, "output_tokens": M } }
//
// Every request is appended to MOCK_LOG (JSONL) as an event the tests assert on.

import { createServer } from "http";
import { appendFileSync } from "fs";

function messageFromTurn(turn, model, reqIndex) {
  const content = [];
  if (turn.text && (turn.tools?.length)) content.push({ type: "text", text: turn.text });
  else if (turn.text) content.push({ type: "text", text: turn.text });
  (turn.tools || []).forEach((t, j) => {
    content.push({ type: "tool_use", id: `toolu_mock_${reqIndex}_${j}`, name: t.name, input: t.input ?? {} });
  });
  const stop_reason = turn.tools?.length ? "tool_use" : "end_turn";
  const usage = turn.usage || { input_tokens: 100, output_tokens: 20 };
  return { id: `msg_mock_${reqIndex}`, type: "message", role: "assistant", model, content, stop_reason, stop_sequence: null, usage };
}

function writeBlocking(res, msg) {
  const body = JSON.stringify(msg);
  res.writeHead(200, { "content-type": "application/json" });
  res.end(body);
}

function sse(res, event, data) {
  res.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
}

async function writeStreaming(res, msg) {
  res.writeHead(200, { "content-type": "text/event-stream", "cache-control": "no-cache", connection: "keep-alive" });
  sse(res, "message_start", { type: "message_start", message: { ...msg, content: [], stop_reason: null, usage: { ...msg.usage, output_tokens: 0 } } });
  msg.content.forEach((block, i) => {
    if (block.type === "text") {
      sse(res, "content_block_start", { type: "content_block_start", index: i, content_block: { type: "text", text: "" } });
      // Split text into a few chunks so streaming is observably chunked.
      const chunks = block.text.match(/.{1,24}(\s|$)|.+$/g) || [block.text];
      for (const c of chunks) sse(res, "content_block_delta", { type: "content_block_delta", index: i, delta: { type: "text_delta", text: c } });
      sse(res, "content_block_stop", { type: "content_block_stop", index: i });
    } else {
      sse(res, "content_block_start", { type: "content_block_start", index: i, content_block: { type: "tool_use", id: block.id, name: block.name, input: {} } });
      sse(res, "content_block_delta", { type: "content_block_delta", index: i, delta: { type: "input_json_delta", partial_json: JSON.stringify(block.input) } });
      sse(res, "content_block_stop", { type: "content_block_stop", index: i });
    }
  });
  sse(res, "message_delta", { type: "message_delta", delta: { stop_reason: msg.stop_reason, stop_sequence: null }, usage: { output_tokens: msg.usage.output_tokens } });
  sse(res, "message_stop", { type: "message_stop" });
  res.end();
}

// Start the mock. Returns { url, close, port }.
// A scenario is either flat ({turns}) — one "main" track — or multi-track
// ({tracks: {main: {turns}, compact: {match, turns}, ...}}). Each request is
// routed to a track by a substring its `match` finds in the system prompt (so an
// aux call like compaction can be told apart from the main loop); "main" is the
// fallback. Each track has its own request counter, so aux calls don't disturb
// the main loop's turn index.
export function startMock({ scenario, logPath } = {}) {
  const tracks = scenario?.tracks || { main: { turns: scenario?.turns || [] } };
  const counters = {};
  let reqIndex = 0;

  const server = createServer((req, res) => {
    if (req.method !== "POST" || !req.url.startsWith("/v1/messages")) {
      res.writeHead(404); res.end("not found"); return;
    }
    let raw = "";
    req.on("data", (c) => (raw += c));
    req.on("end", async () => {
      let body;
      try { body = JSON.parse(raw); } catch { res.writeHead(400); res.end("bad json"); return; }

      const system = typeof body.system === "string" ? body.system : (Array.isArray(body.system) ? body.system.map((b) => b.text).join("") : "");
      const firstUserText = (() => { const m = (body.messages || []).find((x) => x.role === "user"); return typeof m?.content === "string" ? m.content : ""; })();
      const toolNames = (body.tools || []).map((t) => t.name);
      // Route to an aux track by a structured match (system substring + optional
      // firstUser / tools). Check ALL tracks, not first-hit, and fail loudly on
      // ambiguity — a request must never match two tracks (that's a scenario bug,
      // or a track marker string that collided with the main system prompt).
      const matches = (t) =>
        (!t.match || system.includes(t.match)) &&
        (!t.firstUserContains || firstUserText.includes(t.firstUserContains)) &&
        (!t.toolsInclude || t.toolsInclude.every((n) => toolNames.includes(n)));
      const hits = Object.entries(tracks).filter(([name, t]) => name !== "main" && t.match && matches(t));
      let track = "main";
      if (hits.length > 1) {
        const err = { type: "error", error: { type: "mock_ambiguous_route", message: `request matched multiple tracks: ${hits.map(([n]) => n).join(", ")}` } };
        if (logPath) appendFileSync(logPath, JSON.stringify({ type: "ambiguous", req: reqIndex, tracks: hits.map(([n]) => n), system }) + "\n");
        res.writeHead(500, { "content-type": "application/json" }); res.end(JSON.stringify(err)); return;
      }
      if (hits.length === 1) track = hits[0][0];
      const turnIndex = counters[track] || 0;
      const turn = (tracks[track]?.turns || [])[turnIndex];

      // tool_result blocks the agent sent back — proof the tool actually ran,
      // with its real output (a broken tool shows up here as wrong content).
      const toolResults = [];
      for (const m of body.messages || []) {
        if (Array.isArray(m.content)) for (const b of m.content) {
          if (b.type === "tool_result") toolResults.push({ tool_use_id: b.tool_use_id, content: typeof b.content === "string" ? b.content : JSON.stringify(b.content) });
        }
      }
      if (logPath) {
        appendFileSync(logPath, JSON.stringify({
          type: "request",
          req: reqIndex,
          track,
          turnIndex,
          system,
          tools: toolNames,
          toolResults,
          messageCount: (body.messages || []).length,
          firstUserText,
          stream: !!body.stream,
        }) + "\n");
      }

      // Queue exhausted: fail loudly. A false green is worse than a red test.
      if (!turn) {
        const err = { type: "error", error: { type: "mock_exhausted", message: `mock track "${track}" has no turn ${turnIndex}` } };
        if (logPath) appendFileSync(logPath, JSON.stringify({ type: "exhausted", req: reqIndex, track, turnIndex }) + "\n");
        res.writeHead(500, { "content-type": "application/json" }); res.end(JSON.stringify(err)); return;
      }
      counters[track] = turnIndex + 1;

      const msg = messageFromTurn(turn, body.model || "mock", reqIndex);
      if (logPath) appendFileSync(logPath, JSON.stringify({
        type: "response", req: reqIndex, stop_reason: msg.stop_reason,
        tool_use: msg.content.filter((b) => b.type === "tool_use").map((b) => ({ name: b.name, input: b.input })),
      }) + "\n");
      reqIndex++;

      if (body.stream) await writeStreaming(res, msg);
      else writeBlocking(res, msg);
    });
  });

  return new Promise((resolve) => {
    server.listen(0, "127.0.0.1", () => {
      const port = server.address().port;
      resolve({ url: `http://127.0.0.1:${port}`, port, close: () => new Promise((r) => server.close(r)) });
    });
  });
}

File: steps/mock_anthropic.py
"""Python twin of mock-anthropic.mjs. Speaks the real Anthropic /v1/messages
protocol (create + SSE stream, tool_use, usage, errors) and replays a scripted
scenario. Runs in-thread so a Python step in the same process can reach it
(cross-process loopback is intercepted by the dev host's proxy). Reads the same
steps/scenarios/*.json as the Node mock, so the two cannot behave differently.
"""

import json
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer


def _message_from_turn(turn, model, req_index):
    content = []
    if turn.get("text"):
        content.append({"type": "text", "text": turn["text"]})
    for j, t in enumerate(turn.get("tools", [])):
        content.append({"type": "tool_use", "id": f"toolu_mock_{req_index}_{j}",
                        "name": t["name"], "input": t.get("input", {})})
    stop_reason = "tool_use" if turn.get("tools") else "end_turn"
    usage = turn.get("usage", {"input_tokens": 100, "output_tokens": 20})
    return {"id": f"msg_mock_{req_index}", "type": "message", "role": "assistant",
            "model": model, "content": content, "stop_reason": stop_reason,
            "stop_sequence": None, "usage": usage}


def _sse(block_lines):
    return "".join(f"event: {e}\ndata: {json.dumps(d)}\n\n" for e, d in block_lines)


def _stream_body(msg):
    lines = [("message_start", {"type": "message_start", "message": {**msg, "content": [], "stop_reason": None,
              "usage": {**msg["usage"], "output_tokens": 0}}})]
    for i, block in enumerate(msg["content"]):
        if block["type"] == "text":
            lines.append(("content_block_start", {"type": "content_block_start", "index": i,
                          "content_block": {"type": "text", "text": ""}}))
            text = block["text"]
            for k in range(0, len(text), 24):
                lines.append(("content_block_delta", {"type": "content_block_delta", "index": i,
                              "delta": {"type": "text_delta", "text": text[k:k + 24]}}))
            lines.append(("content_block_stop", {"type": "content_block_stop", "index": i}))
        else:
            lines.append(("content_block_start", {"type": "content_block_start", "index": i,
                          "content_block": {"type": "tool_use", "id": block["id"], "name": block["name"], "input": {}}}))
            lines.append(("content_block_delta", {"type": "content_block_delta", "index": i,
                          "delta": {"type": "input_json_delta", "partial_json": json.dumps(block["input"])}}))
            lines.append(("content_block_stop", {"type": "content_block_stop", "index": i}))
    lines.append(("message_delta", {"type": "message_delta",
                  "delta": {"stop_reason": msg["stop_reason"], "stop_sequence": None},
                  "usage": {"output_tokens": msg["usage"]["output_tokens"]}}))
    lines.append(("message_stop", {"type": "message_stop"}))
    return _sse(lines)


def start_mock(scenario, log_path=None):
    """Start the mock in a daemon thread. Returns (url, close). A scenario is
    flat ({turns}) or multi-track ({tracks: {main:{turns}, compact:{match,turns}}});
    requests route to a track by a substring its `match` finds in the system
    prompt (default "main"), each track with its own counter."""
    tracks = (scenario or {}).get("tracks") or {"main": {"turns": (scenario or {}).get("turns", [])}}
    state = {"req": 0, "counters": {}}

    def log(obj):
        if log_path:
            with open(log_path, "a") as f:
                f.write(json.dumps(obj) + "\n")

    class Handler(BaseHTTPRequestHandler):
        def do_POST(self):
            if not self.path.startswith("/v1/messages"):
                self.send_response(404); self.end_headers(); self.wfile.write(b"not found"); return
            body = json.loads(self.rfile.read(int(self.headers.get("content-length", 0))) or b"{}")
            req_index = state["req"]

            sys_text = body.get("system", "")
            if isinstance(sys_text, list):
                sys_text = "".join(b.get("text", "") for b in sys_text)
            tool_names = [t["name"] for t in body.get("tools", [])]
            first_user = next((m for m in body.get("messages", []) if m.get("role") == "user"), None)
            first_user_text = first_user["content"] if first_user and isinstance(first_user.get("content"), str) else ""

            # Route to an aux track by a structured match (system substring +
            # optional firstUser / tools). Check ALL tracks and fail loudly on
            # ambiguity — a request must never match two tracks.
            def _matches(t):
                return ((not t.get("match") or t["match"] in sys_text)
                        and (not t.get("firstUserContains") or t["firstUserContains"] in first_user_text)
                        and (not t.get("toolsInclude") or all(n in tool_names for n in t["toolsInclude"])))
            hits = [name for name, t in tracks.items() if name != "main" and t.get("match") and _matches(t)]
            if len(hits) > 1:
                log({"type": "ambiguous", "req": req_index, "tracks": hits, "system": sys_text})
                err = json.dumps({"type": "error", "error": {"type": "mock_ambiguous_route",
                      "message": f"request matched multiple tracks: {', '.join(hits)}"}}).encode()
                self.send_response(500); self.send_header("content-type", "application/json")
                self.send_header("content-length", str(len(err))); self.end_headers(); self.wfile.write(err); return
            track = hits[0] if len(hits) == 1 else "main"
            turn_index = state["counters"].get(track, 0)
            track_turns = tracks.get(track, {}).get("turns", [])
            turn = track_turns[turn_index] if turn_index < len(track_turns) else None

            # tool_result blocks the agent sent back — proof the tool actually
            # ran, with its real output.
            tool_results = []
            for m in body.get("messages", []):
                if isinstance(m.get("content"), list):
                    for b in m["content"]:
                        if b.get("type") == "tool_result":
                            c = b.get("content")
                            tool_results.append({"tool_use_id": b.get("tool_use_id"),
                                                 "content": c if isinstance(c, str) else json.dumps(c)})
            log({"type": "request", "req": req_index, "track": track, "turnIndex": turn_index,
                 "system": sys_text, "tools": tool_names,
                 "toolResults": tool_results, "messageCount": len(body.get("messages", [])),
                 "firstUserText": first_user_text, "stream": bool(body.get("stream"))})

            if turn is None:
                log({"type": "exhausted", "req": req_index, "track": track, "turnIndex": turn_index})
                err = json.dumps({"type": "error", "error": {"type": "mock_exhausted",
                      "message": f'track "{track}" has no turn {turn_index}'}}).encode()
                self.send_response(500); self.send_header("content-type", "application/json")
                self.send_header("content-length", str(len(err))); self.end_headers(); self.wfile.write(err); return
            state["counters"][track] = turn_index + 1

            msg = _message_from_turn(turn, body.get("model", "mock"), req_index)
            log({"type": "response", "req": req_index, "stop_reason": msg["stop_reason"],
                 "tool_use": [{"name": b["name"], "input": b["input"]} for b in msg["content"] if b["type"] == "tool_use"]})
            state["req"] += 1

            if body.get("stream"):
                b = _stream_body(msg).encode()
                self.send_response(200); self.send_header("content-type", "text/event-stream")
                self.send_header("content-length", str(len(b))); self.end_headers(); self.wfile.write(b)
            else:
                b = json.dumps(msg).encode()
                self.send_response(200); self.send_header("content-type", "application/json")
                self.send_header("content-length", str(len(b))); self.end_headers(); self.wfile.write(b)

        def log_message(self, *a):
            pass

    srv = HTTPServer(("127.0.0.1", 0), Handler)
    port = srv.server_address[1]
    threading.Thread(target=srv.serve_forever, daemon=True).start()
    return f"http://127.0.0.1:{port}", srv.shutdown
